### PR TITLE
 free ends alignment configuration

### PR DESCRIPTION
**Pull request by joergi-w**
solves issues #335 and #336

### CODE REVIEW

File: include/seqan3/alignment/configuration/align_config_free_ends.hpp [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+enum struct free_ends_enum
+{
+    none       = 0,
+    seq1_front = 1 << 0,
+    seq1_back  = 1 << 1,
+    seq2_front = 1 << 2,
+    seq2_back  = 1 << 3,
+    seq1       = seq1_front | seq1_back,
+    seq2       = seq2_front | seq2_back,
+    front      = seq1_front | seq2_front,
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_free_ends.hpp]: I would remove the front and back shortcuts.

File: include/seqan3/alignment/configuration/align_config_free_ends.hpp [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+{
+    //! \brief Holds the actual free ends settings.
+    free_ends_enum value = pos;
+};
+
+/*!\brief The free ends adaptor enabling pipe notation.
+ * \ingroup configuration
+ * \tparam free_ends_cfg_t A template-template class specifying the actual free ends settings.
+ */
+template <template <free_ends_enum> typename free_ends_cfg_t, free_ends_enum e>
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_free_ends.hpp]: why this superfluous template template class

File: include/seqan3/alignment/configuration/align_config_free_ends.hpp [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    template <typename configuration_type>
+    //!\cond
+        requires is_algorithm_configuration_v<remove_cvref_t<configuration_type>>
+    //!\endcond
+    constexpr auto invoke(configuration_type && cfg, free_ends_enum const fe) const
+    {
+        static_assert(is_valid_alignment_configuration_v<align_cfg::id::free_ends, remove_cvref_t<configuration_type>>,
+                      SEQAN3_INVALID_CONFIG(align_cfg::id::free_ends));
+
+        return std::forward<configuration_type>(cfg).push_front(align_config_free_ends<>{std::move(fe)});
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_free_ends.hpp]: You need for this the deferred config class. Otherwise, the value represented does not represent the template argument of the config. That could lead to misuse in the code.

File: include/seqan3/alignment/configuration/utility.hpp [hunk 4]
```diff
@@ -66,7 +66,17 @@ namespace seqan3::align_cfg
...
-    gap
+    gap,
+    //!\brief Identifier for band configuration.
+    band,
+    //!\brief Identifier for free ends configuration.
+    free_ends,
+    //!\brief Identifier for score configuration.
+    score,
+    //!\brief Identifier for edit distance configuration.
+    edit,
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/utility.hpp]: We do not need an extra field for edit. 
edit will be a combination of known variables already.

File: include/seqan3/alignment/configuration/align_config_free_ends.hpp [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+/*!\brief A configuration adaptor for gaps.
+ * \ingroup configuration
+ * \tparam pos The compile time value for this free ends configuration.
+ *
+ * \details This configuration allows to specify, whether continuous gaps in the front or end of a sequence
+ * are penalized in the alignment.
+ */
+template <free_ends_enum pos = free_ends_enum::none>
+inline constexpr detail::align_config_free_ends_adaptor<seqan3::detail::align_config_free_ends, pos> free_ends;
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_free_ends.hpp]: What do you think about renaming the config to:
```cpp
align_cfg::sequence_ends<free_end_at::seq1_front | free_end_at::seq1_back>
```

File: include/seqan3/alignment/configuration/utility.hpp (hunk 4 above)
Comment (joergi-w) [reply include/seqan3/alignment/configuration/utility.hpp]: Ok!

**Comment by rrahn** [comment]
The doc devel test failed:
```
1: b'Parsing file /h/home/jenkins/ws/pipeline_doc_tests/bb70ec0a/checkout-31d65c68d639ff93adef914ab5844bcbc33e8820/include/seqan3/alignment/configuration/align_config_free_ends.hpp:229: warning: Compound seqan3::detail::align_config_type_to_id< align_config_free_ends_deferred > is not documented.\n'
1: b"/home/jenkins/ws/pipeline_doc_tests/bb70ec0a/checkout-31d65c68d639ff93adef914ab5844bcbc33e8820/include/seqan3/alignment/configuration/align_config_free_ends.hpp:172: warning: argument 'fe' of command @param is not found in the argument list of seqan3::detail::align_config_free_ends_adaptor< e >::invoke(configuration_type &&cfg, free_ends_at const _e) const\n"
1: b'/home/jenkins/ws/pipeline_doc_tests/bb70ec0a/checkout-31d65c68d639ff93adef914ab5844bcbc33e8820/include/seqan3/alignment/configuration/align_config_free_ends.hpp:175: warning: The following parameters of seqan3::detail::align_config_free_ends_adaptor::invoke(configuration_type &&cfg, free_ends_at const _e) const are not documented:\n'
1: b'/home/jenkins/ws/pipeline_doc_tests/bb70ec0a/checkout-31d65c68d639ff93adef914ab5844bcbc33e8820/include/seqan3/alignment/configuration/align_config_free_ends.hpp:100: warning: Member invoke(fn_t &&fn, configuration_t &&cfg) const requires detail (function) of class seqan3::detail::align_config_free_ends_deferred is not documented.\n'
1: b'/home/jenkins/ws/pipeline_doc_tests/bb70ec0a/checkout-31d65c68d639ff93adef914ab5844bcbc33e8820/include/seqan3/alignment/configuration/align_config_free_ends.hpp:97: warning: Member value (variable) of class seqan3::detail::align_config_free_ends_deferred is not documented.\n'
1: Detected 5 warnings!
1/1 Test #1: test_devel_doc ...................***Failed   12.65 sec
```

File: include/seqan3/alignment/configuration/align_config_free_ends.hpp [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+#include <seqan3/core/metafunction/template_inspection.hpp>
+
+namespace seqan3
+{
+
+/*!\brief Enum class for all supported free ends specifications.
+ * \ingroup alignment_configuration
+ * \details The members specify where continuous gaps in the front or end of a sequence are not penalized
+ * in the alignment.
+ */
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_free_ends.hpp]: The enum values should be documented.

File: include/seqan3/alignment/configuration/utility.hpp [hunk 7]
```diff
@@ -67,8 +67,16 @@ enum struct id : uint8_t
...
+    band,
+    //!\brief Identifier for free ends configuration.
+    sequence_ends,
+    //!\brief Identifier for score configuration.
+    score,
     //!\brief Identifier for global alignment configuration.
-    global
+    global,
+    //!\brief Identifier for output configuration.
+    output
```
Comment (marehr) [inline include/seqan3/alignment/configuration/utility.hpp]: This is not part of the PR, right?

Comment (joergi-w) [reply include/seqan3/alignment/configuration/utility.hpp]: True, but I made one commit (fc8e3f) that already includes all config ids. So I can base all configs on that commit, such that we can individually merge them later without conflicts in this file.

Comment (rrahn) [reply include/seqan3/alignment/configuration/utility.hpp]: Could you remove the band identifier from this commit, and then rebase all PRs?
The band won't be there in a while.

**Comment by joergi-w** [comment]
Sure, it's done!
